#878 - update legacy ResEvap Documentation (#910)
opendcs/opendcs: docs/source/legacy-resevap-computation.rst

Author: Text cap is complete, need to complete for long equations without causing errors with :math: syntax

Author: updates completed

.. _legacy-resevap-computation:

####################################
W9123819P0073–2 ResEvap Computation
####################################

> *Author an engineering and software requirements document detailing ResEvap*

Author: I edited the sentence and will push change with column cap

In order to improve the usability of ResEvap throughout the USACE, the model was
converted to Java and Jython/Python by RMA. At the time of the original conversion,
the desired outcome was a direct translation of the computation logic from
Fortran to Java with Jython/Python scripting used to orchestrate the computations.
Since that time, a need for documentation of the computational routines within
ResEvap has developed, motivating the production of this document.

This documentation is intended to describe both the core computational
functionality, as well as the entry points to the computational routines
currently used by the USACE. The existing entry point for the ResEvap
computations is the Jython Processor, described in the next section. Next, the
computations, including the evaporation, radiation and vertical temperature
profile models are described. Descriptions of the computations are supported
in `Appendix 1`_ and `Appendix 2`_, which contain variable definitions and
constant value listings. `Appendix 3`_ provides example input files.
`Appendix 4`_ provides recommendations for the existing ResEvap model,

> and last there is a list of references.

Author: @rma-psmorris

> ResEvap Software Distribution and USAGE

Reviewer: ```suggestion
ResEvap Software Distribution and Usage
```

Having usage in all caps was confusing, at first I thought perhaps that should've been USACE.

Author: this section has been removed from the documentation.

> ResEvap Software Distribution and USAGE

Reviewer 2: I agree I see no reason for it to be included in the final documentation.

> ResEvap Software Distribution and USAGE

Reviewer 3: That is correct, please remove entire section.

should have an identical name to the RES python argument. This file contains the
latitude, longitude, time zone, rating curve, and input data time series
identifiers. Although the file contains latitude, longitude and time zone
information, these parameters are overridden by the location information in the
CWMS database. The current release of ResEvap requires that there be a
synchronization between the maximum temperature depth location level, the number
of water temperature profile time series, and the elevation extents of the
elevation;area rating curve. The maximum temperature depth determines the number
of required water temperature profiles. The elevation;area curve’s elevation
extents must cover the all of the valid elevations that might occur given
variance in pool elevation. This must currently be manually configured through
setting the level value, establishing and identifiying the profile time series
in the database and the reservoir file, then editing the database rating curve
as needed. A recommendation has been made to calculate the number water
temperature profile time series using the depth level, this is covered in
`Appendix 4`_: Recommendations.

> ResEvap Jython Sequence of Operations

Author: @jbatterman Please verify what you would like to be updated and/or removed for this section: ResEvap Jython Sequence of Operations.

customizable. The parameter, parameter type, time step, and units are all
pre-determined within ResEvap. The output time series include hourly water
surface temperature, sensible heat flux, latent heat flux, solar radiation,
downwelling longwave radiation, upwelling longwave radiation, and evaporation,
daily evaporation, evaporation as flow, and daily water profile temperatures.
Note that the water temperature profile values are computed at an hourly time
step, but only saved to the CWMS database as a daily time series. After storing
these time series to the CWMS database, the database connection is closed, and
the script is completed.

Java Compute Routines
=====================

Core Routines
--------------

The Java entry point is a class called ResEvap. This class is tasked with
building the necessary compute objects, looping through each hourly time step,
organizing the results, and creating legacy text-based results files.
The ResEvap compute object is constructed by the Jython processor, and
subsequently provided with meteorological data and reservoir information.
Based on this information, ResEvap runs the compute, at an hourly time-step,

> and then writes the output time series to text files.

Reviewer 2: The only key differences is that the ResEvap Java class is now constructed by the ResEvapAlgo class within OpenDCS which constructs it using the parameters set and saved by the user for the algorithm within their OpenDCS database. Also, as you metioned there is no reading or writng of data to files from the ResEvap class now all data is returned to the ResEvapAlgo class which then stores and returns it to the OpenDCS database or a logger set by the user.

> and then writes the output time series to text files.

Author: @jbatterman do you want to rewrite this section based on the last comment from Oskar?

ResEvap assumes that the vertical temperature profile is governed by the energy
transfer through the water surface, which requires that the inflows and outflows
are negligible components of the energy balance. This is appropriate for the
Missouri River watershed :ref:`(Daly 2005) <References>`, which ResEvap was
designed for, but this assumption should be carefully analyzed when applying to
new watersheds. Four heat transfer modes are computed: sensible, latent,
shortwave radiation and longwave radiation. Computation of these fluxes are
described in the “Evaporation Computations” and “Radiation Computations” sections
of this document. Based on these fluxes, and the initial water temperature at
each layer in the profile, the temperatures are updated as described in the
“Vertical Temperature Profile Computations” section.

**Note:** `Appendix 1`_ **has a definition of all variables in the computations,
and** `Appendix 2`_ **has a definition of all the constants used.**

Input Data
~~~~~~~~~~

Input data for ResEvap includes a combination of meteorological observations and
reservoir physical parameters. The time series observations are provided in time

> series form, and are summarized in Table 2. In this table, note that the cloud
> heights are estimated by ResEvap if not provided, making them an optional
> parameter.

Reviewer 3: ```suggestion
series form, and are summarized in Table 2. In this table, note that the cloud
heights are estimated by ResEvap if not provided, making them an optional
parameter. The input time series is required but the computation will still
calculate if there is missing data.
```

.. _Table 2:

> *Table 2- Time series input data required for ResEvap computations*

Reviewer 3: @oskarhurst will the computation break if the optional cloud height time series are not defined?  Or do they have to be defined but can have missing data?

| High Cloud Fraction            | Instantaneous        | 1Hour         | .. math:: \%          |
+--------------------------------+----------------------+---------------+-----------------------+
| High Cloud Height              | Instantaneous        | 1Hour         | .. math:: m           |
+--------------------------------+----------------------+---------------+-----------------------+
| Reservoir Pool Elevation       | Instantaneous        | 1Hour         | .. math:: m           |
+--------------------------------+----------------------+---------------+-----------------------+
| Water Temperature (Each Layer) | Instantaneous        | 1Day          | .. math:: {^\circ}C   |
+--------------------------------+----------------------+---------------+-----------------------+

In addition to time series data, ResEvap requires the start date, end date,
GMT offset, version name, latitude, longitude, observation heights for wind
speed, air temperature and relative humidity, and the elevation-area rating
curve. Note that ResEvap is not aware of vertical datum info. All elevation
input data must be supplied in the same vertical datum.

Output Data
~~~~~~~~~~~

> ResEvap produces both meteorological and water temperature information, which is
> written to text files and returned to the Jython processor for storage into the
> CWMS database. Table 3 summarizes the time series data produced by ResEvap.

Reviewer 3: ```suggestion
ResEvap produces both meteorological and water temperature information which is
returned to the ResEvapAlgo class for storage into the CWMS database. Table 3 
summarizes the time series data produced by ResEvap.
```

+--------------------------------+---------------------------+---------------+---------------------------+
| Sensible Heat Flux             | Instantaneous             | 1Hour         | .. math:: \frac{W}{m^{2}} |
+--------------------------------+---------------------------+---------------+---------------------------+
| Latent Heat Flux               | Instantaneous             | 1Hour         | .. math:: \frac{W}{m^{2}} |
+--------------------------------+---------------------------+---------------+---------------------------+
| Evaporation                    | Instantaneous, Cumulative | 1Hour, 1Day   | .. math:: mm              |
+--------------------------------+---------------------------+---------------+---------------------------+
| Evaporation as Flow            | Average                   | 1Day          | .. math:: \frac{m^{3}}{s} |
+--------------------------------+---------------------------+---------------+---------------------------+
| Water Temperature (Each Layer) | Instantaneous             | 1Hour         | .. math:: {^\circ}C       |
+--------------------------------+---------------------------+---------------+---------------------------+

> ResEvap builds these output time series based on the input time window, location
> and version name. As the compute progresses in time, the hourly time series are
> filled with compute results. At the end of the simulation, the evaporation is
> accumulated to daily, and then the evaporation as flow is computed from the
> daily evaporation, and the reservoir surface area. This is converted based on
> the following equation:

Reviewer 3: ```suggestion
ResEvap builds these output time series based on the location and version name. 
As the compute progresses in time, the hourly time series are filled with compute 
results. At the end of the simulation, the evaporation is accumulated to daily, 
and then the evaporation as flow is computed from the daily evaporation and the 
reservoir surface area. This is converted based on the following equation:
```

:math:`{E_{f_t}} = E_{t}{A_{s_t}}`

Where :math:`E_{f_t}` is the evaporation as flow at time :math:`t`, :math:`E_{t}`
is the evaporation rate at time :math:`t`, and :math:`A_{s_t}` is the reservoir
surface area at time :math:`t`.

Time series data is saved to text files as well. Time series data is reported in
text files named "testtout_java.dat", "wtout_java.dat", reporting the
meteorological/surface flux and water temperature time series information,
respectively. The meteorological/surface flux results file reports values at
every hour for the wind speed, air temperature, relative humidity, air pressure,
water surface temperature, :math:`u_{*}`, :math:`{R_{e}}^{*}`, Obhukov Length,
sensible heat flux, latent heat flux, solar radiation flux, downwelling longwave
radiation flux, upwelling longwave radiation flux, and evaporation. The water
temperature results file reports the temperature of every layer, for every hour

> computed, in :math:`℃`.

Reviewer 3: Remove this section, we no longer have any results files.

ResEvap also saves diagnostic information to text files, which can be used for
debugging purposes. Reservoir profile and energy balance reports are provided in
files named "xout_java.dat" and "xout2_java.dat", respectively. The reservoir
profile information includes the depth to each layer, the thickness of each
layer, the area of each layer, the elevation of each layer, and the volume of
each layer. The energy balance report contains the water surface elevation,
total thermal energy, the change in total thermal energy, the total thermal
energy input, the total thermal energy at the end of the time step, the relative
difference between the change in thermal energy and the total (net) energy input
(should be ~0), and the reservoir surface area.

A recommendation has been added to `Appendix 4`_: Recommendations, to prevent

> the generation of these files unless ResEvap is being run in a debug mode.

Reviewer 3: Recommend removing these paragraphs as well which discuss results files.

Evaporation Computations
------------------------

> Evaporation computations are performed in the EvapWater class. The evaporation

Reviewer 2: Yes this is still true the majority of the computation is still preformed in the EvapWater class. This Class call mutliple other classes which compute the sub values nessicary for solving for resivoir Evaporation. An important note for changes In the ResEvap class is it now used the OpenDCs TimeSeires object to handle storing and data manipulation of timeseries data, as well as it now used the CWMS rating table library to better calcuate rating results which are used in the calculation.

> Evaporation computations are performed in the EvapWater class. The evaporation

Author: @jbatterman Please let me know if you would like to update this section with any information Oskar provided.

> Evaporation computations are performed in the EvapWater class. The evaporation

Reviewer 3: No further modifications to this section with that information.  I will look above and add additional review suggestions based on this note.

*Table 13 - Cloud height coefficients for non-winter season and latitude > 25*

+--------------+--------------+--------------+--------------+---------------+
|              | a            | b            | c            | d             |
+==============+==============+==============+==============+===============+
| Low          | 1.15         | 0.6          | 1.5          | 25.0          |
+--------------+--------------+--------------+--------------+---------------+
| Mid          | 4.4          | 1.2          | 3.0          | 25.0          |
+--------------+--------------+--------------+--------------+---------------+
| High         | 7.0          | 1.5          | 3.0          | 30.0          |
+--------------+--------------+--------------+--------------+---------------+

Vertical Temperature Profile Computations
-----------------------------------------

Vertical transfer of heat within a reservoir is assumed to be a one-dimensional
process, where the reservoir is assumed to be laterally homogeneous. This allows
for ignoring effects of reservoir inflows and outflows. In the event that there
is a large lateral variation in temperature (i.e. long run-of-the-river reservoirs),

> these computations will be unreliable. Based on this assumption, vertical transfer

Author: I spoke with Caleb whose notes reflected that 30 days was his previous recommendation for flushing time, I have updated this section with his advisement.

stable. At this point, it is still possible deeper layers are in the SML, due to
the combined convective and wind driven turbulent energy. Therefore, the
turbulent kinetic energy (TKE) must be computed, and compared against the
potential energy.

:math:`{TKE}_{i:N} = Ke_{c_{i:N}} + Ke_{u_{i:N}}`

:math:`Ke_{c_{i:N}} = \dfrac{\varepsilon_{c}g}{\rho_{N}\mathrm{\Delta}t} \biggl( \sum_{k = i}^{N} \Bigl( \rho_{k} \
( z_{k} - z_{k - 1} ) \frac{( z_{k} + z_{k - 1} )}{2} \Bigr) - \frac{( z_{N} + z_{i - 1} )}{2} \sum_{k = i}^{N} \
\Bigl( \rho_{k}( z_{k} - z_{k - 1} ) \Bigr) \biggr)`

:math:`Ke_{u_{i:N}} = \varepsilon_{u}\rho_{N}A_{N}{u_{*}^{w}}^{3}\mathrm{\Delta}t`

:math:`u_{*}^{w} = u_{*}\sqrt{\frac{\rho_{a}}{\rho_{N}}}`

Where :math:`Ke_{c_{i:N}}` is the kinetic energy of the SML with layer
:math:`i` included and :math:`Ke_{u_{i:N}}` is the kinetic energy from wind
with layer :math:`i` included. If :math:`TKE_{i:N} \geq PE_{mix_i}` , then layer
:math:`i` is considered in the SML, and the computations checks the deeper layer.
If :math:`TKE_{i:N} > PE_{mix_i}` , then the computation of vertical
temperature profile is complete. At this point, the reservoir surface

> temperature computations have comple ted, and ResEvap moves on to the next
> time step. After the final time step, ResEvap reports data in the output reports
> and returns the results to the Jython processor.

Reviewer 2: ```suggestion
temperature computations have completed, and ResEvap moves on to the next
time step. After each time step the parliminarly caclualted values for that hour are then
returned to the ResEvapAlgo class which preps the results to be stored in the users 
OpenDCS database of choice. The ResEvapAlgo class then calls the ResEvap class to cumputed
the results for the next time step. Once all time steps are calculated for a span of one day
24 hours (23 and 25 for dayligth savings time) Daliy average and total values for Evaporation
and water profile tempetures are calculated. Once the entire aggregating period is solved for
the ResEvapAlgo class then executes the write action to store all values to the users database.
```

> and returns the results to the Jython processor.

Author: Note: resolved in a comment below.

+---------------------------+------------------------------+-------------------------+--------------------------------+
| .. math:: \kappa          | von Karman constant          | .. math:: 0.4           | unitless                       |
+---------------------------+------------------------------+-------------------------+--------------------------------+
| .. math:: \sigma          | Stefan-Boltzman constant     | .. math:: 5.67*10^{- 8} | .. math:: \frac{W}{m^{2}K^{4}} |
+---------------------------+------------------------------+-------------------------+--------------------------------+
| .. math:: \theta          | theta method factor          | .. math:: 1             | unitless                       |
+---------------------------+------------------------------+-------------------------+--------------------------------+

NOTE: The Stefan-Boltzman constant is :math:`5.669*10^{- 8}` in the computation
of the incoming longwave radiation, which is slightly different than the rest
of the computations. This is considered an insignificant difference.

.. _Appendix 3:

> Appendix 3: Example Input Files

Author: I have created a new ticket to verify these items in Appendix 3 are in @oskarhurst documentation.

    53:SWT:PA18.Temp-Water-11,5m.Inst.1Day.0.nwo-evaptest:C
    54:SWT:PA18.Temp-Water-12,0m.Inst.1Day.0.nwo-evaptest:C
    55:SWT:PA18.Temp-Water-12,5m.Inst.1Day.0.nwo-evaptest:C
    56:SWT:PA18.Temp-Water-13,0m.Inst.1Day.0.nwo-evaptest:C
    57:SWT:PA18.Temp-Water-13,5m.Inst.1Day.0.nwo-evaptest:C
    58:SWT:PA18.Temp-Water-14,0m.Inst.1Day.0.nwo-evaptest:C
    59:SWT:PA18.Temp-Water-14,5m.Inst.1Day.0.nwo-evaptest:C
    60:SWT:PA18.Temp-Water-15,0m.Inst.1Day.0.nwo-evaptest:C
    61:SWT:PA18.Temp-Water-15,5m.Inst.1Day.0.nwo-evaptest:C
    62:SWT:PA18.Temp-Water-16,0m.Inst.1Day.0.nwo-evaptest:C
    63:SWT:PA18.Temp-Water-16,5m.Inst.1Day.0.nwo-evaptest:C
    64:SWT:PA18.Temp-Water-17,0m.Inst.1Day.0.nwo-evaptest:C
    65:SWT:PA18.Temp-Water-17,5m.Inst.1Day.0.nwo-evaptest:C
    66:SWT:PA18.Temp-Water-18,0m.Inst.1Day.0.nwo-evaptest:C
    67:SWT:PA18.Temp-Water-18,5m.Inst.1Day.0.nwo-evaptest:C
    68:SWT:PA18.Temp-Water-19,0m.Inst.1Day.0.nwo-evaptest:C
    69:SWT:PA18.Temp-Water-19,5m.Inst.1Day.0.nwo-evaptest:C
    70:SWT:PA18.Temp-Water-20,0m.Inst.1Day.0.nwo-evaptest:C
    71:SWT:PA18.Temp-Water-20,5m.Inst.1Day.0.nwo-evaptest:C

.. _Appendix 4:

> Appendix 4: Recommendations

Author: I will remove this pending approved removal of sections and/or recommendation comments that reference Appendix 4.

> Appendix 4: Recommendations

Author: @jbatterman, @rma-psmorris per our conversation regarding Appendix 4:

**Delete (no longer relevant or completed):**
#2 Simplify ResEvap input/output variable configuration
#5 Change program Name
#12 Re-code the ResEvap shell script as a bash script
#13 Cleanup unused file production

**Create a Ticket:**
#1 Implement Automated Testing 
#3 ResEvap initialization process
#8 Add frustum computation (ticket #1023)
#9 Add support for solar radiation observations
#10 Fix vertically averaged density computations (ticket #1025)
#14 Allow for use of fog/smoke layer computations

**Should Capture in Confluence (delete from documentation):**
#4 Compare ResEvap against other water temperature models
#6 Add additional user configuration

**Pending Verification from Oskar and Jessica:**
#11 Allow for storage of hourly temperature profiles

**Needs to be in Documentation:**
#7 Add vertical datum support
The user should make sure the elevation readings are in the same datum as the rating curve.  Might actually be an error and not an edge case. NOTE needs to be in Document: make sure the datums are the same.

@jbatterman, please let me know if/when you would like Appendix 4 removed from documentation completely
